src/python/pants/engine/internals/scheduler.py

```diff
@@ -614,11 +614,23 @@ def merge_directories(self, digests):
             self._scheduler._scheduler, self._session, _DirectoryDigests(digests)
         )
 
-    def digests_to_bytes(self, digests: Sequence[Digest]) -> Tuple[bytes]:
+    def single_file_digests_to_bytes(self, digests: Sequence[Digest]) -> Tuple[bytes]:
         sched_pointer = self._scheduler._scheduler
         return cast(
             Tuple[bytes],
-            tuple(self._scheduler._native.lib.digests_to_bytes(sched_pointer, list(digests))),
+            tuple(
+                self._scheduler._native.lib.single_file_digests_to_bytes(
+                    sched_pointer, list(digests)
+                )
+            ),
+        )
+
+    def snapshots_to_file_contents(
+        self, snapshots: Sequence[Snapshot]
+    ) -> Tuple[DigestContents, ...]:
+        return tuple(
+            cast(DigestContents, self.product_request(DigestContents, [snapshot.digest]))
```

Comment:
I think you can parallelize this:

```python
return tuple(self.product_request(DigestContents, [snapshot.digest for snapshot in snapshots])
```

The previous implementation is the equivalent of `Get` in a `for` loop. This rewrite is like `MultiGet`. (I think - have to run to a meeting so can't verify)

Reply:
I think this requests one `DigestContents` object out of the entire collection of `Snapshot`s; but we want to request one `DigestContents` *per* `Snapshot` (the idea is that this method can be called once with as many artifact snapshots as a plugin needs to handle).

Reply:
Eric is right I think: `product_request` supports getting the same output type for multiple inputs in parallel. The API is definitely confusing, sorry.

Reply:
Talked with @Eric-Arellano about this. Using the parallel API would end up smashing `FileContent`s from different `Snapshot`s representing different artifacts together, which we don't want; I'm adding some comments and modifying the tests to clarify this.